### PR TITLE
[new release] caisar, caisar-ovo, caisar-onnx, caisar-nnet, caisar-xgboost and caisar-ir (0.2)

### PR DESCRIPTION
**Pull request by caisar-platform**
A platform for characterizing the safety and robustness of artificial intelligence based software

- Project page: <a href="https://git.frama-c.com/pub/caisar">https://git.frama-c.com/pub/caisar</a>
- Documentation: <a href="https://git.frama-c.com/pub/caisar">https://git.frama-c.com/pub/caisar</a>

##### CHANGES:

- [prover] Integration of the [nnenum](https://github.com/stanleybak/nnenum)
  prover.

- [prover] Integration of the
  [$\alpha-\beta-$CROWN](https://github.com/stanleybak/nnenum) prover.

- [prover] Integration of the AIMOS metamorphic testing prover.

- [prover] Add printer for VNN-LIB format for property specification as
  supported by nnenum and $\alpha-\beta-$CROWN provers.

- [prover] Support for multiple configurations of provers, using the
  prover-altern command line option. An example of registering an alternate
  configuration is available under config/caisar-detection.conf

- [prover] Add transformation to translate ONNX format into SMTLIB format for
  allowing SMTLIB2 compliant provers to work on neural networks.

- [doc] The first version of the CAISAR manual is publicly available on our
  [website](caisar-platform.com/documentation). It includes detailed
  installation instructions, examples on the ACAS-Xu and MNIST local robustness
  benchmarks. Those examples use an experimental interpretation language that is
  not fully documented yet.

- [deps] Upgraded Why3 to 1.6.0 version.

- Add verify-json command for verifying a robustness property via a JSON
  configuration file.

- Add verification of datasets for classification tasks in terms of a specific
  CSV format: each line provides the label in 1st column, and data features in
  the other columns.

- Add support for memory and time limits.

- Add debug logging options.

### CODE REVIEW

File: packages/caisar-onnx/caisar-onnx.0.2/opam [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+license: "LGPL-2.1-only"
+homepage: "https://git.frama-c.com/pub/caisar"
+doc: "https://git.frama-c.com/pub/caisar"
+bug-reports: "https://git.frama-c.com/pub/caisar/issues"
+depends: [
+  "dune" {>= "2.9"}
+  "ocaml" {>= "4.13"}
+  "base" {>= "v0.14.0"}
+  "stdio" {>= "v0.14.0"}
+  "ocaml-protoc-plugin" {>= "4.2.0"}
```
Comment (mseri) [inline packages/caisar-onnx/caisar-onnx.0.2/opam:16]:
```suggestion
  "caisar-ir" {=version}
  "ocaml-protoc-plugin" {>= "4.2.0"}
```
Due to
```

#=== ERROR while compiling caisar-onnx.0.2 ====================================#
# context              2.2.0~alpha~dev | linux/x86_64 | ocaml-base-compiler.4.14.1 | pinned(https://git.frama-c.com/api/v4/projects/1082/packages/generic/caisar/0.2/caisar-0.2.tbz)
# path                 ~/.opam/4.14/.opam-switch/build/caisar-onnx.0.2
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p caisar-onnx -j 71 --promote-install-files=false @install
# exit-code            1
# env-file             ~/.opam/log/caisar-onnx-7-439e8c.env
# output-file          ~/.opam/log/caisar-onnx-7-439e8c.out
### output ###
# File "lib/onnx/dune", line 4, characters 57-66:
# 4 |  (libraries base stdio ocaml-protoc-plugin ocplib-endian caisar-ir)
#                                                              ^^^^^^^^^
# Error: Library "caisar-ir" not found.
# -> required by library "caisar-onnx" in _build/default/lib/onnx
# -> required by _build/default/META.caisar-onnx
# -> required by _build/install/default/lib/caisar-onnx/META
# -> required by _build/default/caisar-onnx.install
# -> required by alias install
```

File: packages/caisar-ir/caisar-ir.0.2/opam [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+  "LAISER team, Software Safety and Security Laboratory, CEA-List"
+]
+authors: ["LAISER team, Software Safety and Security Laboratory, CEA-List"]
+license: "LGPL-2.1-only"
+homepage: "https://git.frama-c.com/pub/caisar"
+doc: "https://git.frama-c.com/pub/caisar"
+bug-reports: "https://git.frama-c.com/pub/caisar/issues"
+depends: [
+  "ocaml" {>= "4.13"}
+  "dune" {>= "2.9" & >= "2.9.3"}
```
Comment (mseri) [inline packages/caisar-ir/caisar-ir.0.2/opam:13]:
```suggestion
  "dune" {>= "2.9.3"}
```
(This is fixed by removing dune as a dependency in the dune file, dune adds itself automatically. In this case it would add 2.9 probably though)

File: packages/caisar-ir/caisar-ir.0.2/opam [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+  "base" {>= "v0.14.0"}
+  "ocamlgraph" {>= "1.8.8"}
+  "ppx_inline_test" {>= "0.12.0"}
+  "ppx_deriving" {>= "4.4.1"}
+  "odoc" {with-doc}
```
Comment (mseri) [inline packages/caisar-ir/caisar-ir.0.2/opam:18]:
```suggestion
  "caisar-xgboost" {=version & with-test}
  "odoc" {with-doc}
```
Due to
```
#=== ERROR while compiling caisar-ir.0.2 ======================================#
# context              2.2.0~alpha~dev | linux/x86_64 | ocaml-base-compiler.4.14.1 | pinned(https://git.frama-c.com/api/v4/projects/1082/packages/generic/caisar/0.2/caisar-0.2.tbz)
# path                 ~/.opam/4.14/.opam-switch/build/caisar-ir.0.2
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p caisar-ir -j 127 --promote-install-files=false @install @runtest
# exit-code            1
# env-file             ~/.opam/log/caisar-ir-7-074d83.env
# output-file          ~/.opam/log/caisar-ir-7-074d83.out
### output ###
# File "lib/xgboost/bin/dune", line 3, characters 34-37:
# 3 |  (libraries yojson caisar_xgboost csv))
#                                       ^^^
# Error: Library "csv" not found
```

File: packages/caisar-nnet/caisar-nnet.0.2/opam [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+license: "LGPL-2.1-only"
+homepage: "https://git.frama-c.com/pub/caisar"
+doc: "https://git.frama-c.com/pub/caisar"
+bug-reports: "https://git.frama-c.com/pub/caisar/issues"
+depends: [
+  "dune" {>= "2.9"}
+  "ocaml" {>= "4.13"}
+  "base" {>= "v0.14.0"}
+  "csv" {>= "2.4"}
+  "odoc" {with-doc}
```
Comment (mseri) [inline packages/caisar-nnet/caisar-nnet.0.2/opam:16]:
```suggestion
  "caisar-xgboost" {=version & with-test}
  "odoc" {with-doc}
```
Due to
```
#=== ERROR while compiling caisar-nnet.0.2 ====================================#
# context              2.2.0~alpha~dev | linux/x86_64 | ocaml-base-compiler.4.14.1 | pinned(https://git.frama-c.com/api/v4/projects/1082/packages/generic/caisar/0.2/caisar-0.2.tbz)
# path                 ~/.opam/4.14/.opam-switch/build/caisar-nnet.0.2
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p caisar-nnet -j 71 --promote-install-files=false @install @runtest
# exit-code            1
# env-file             ~/.opam/log/caisar-nnet-7-d857a6.env
# output-file          ~/.opam/log/caisar-nnet-7-d857a6.out
### output ###
# File "lib/xgboost/bin/dune", line 3, characters 19-33:
# 3 |  (libraries yojson caisar_xgboost csv))
#                        ^^^^^^^^^^^^^^
# Error: Library "caisar_xgboost" not found.
```

**Comment by mseri** [comment]
caisar-xgboost tests fail on all machines that on,y have the bytecode compiler. Do you maybe understand why?

**Comment by caisar-platform** [comment]
> caisar-xgboost tests fail on all machines that on,y have the bytecode compiler. Do you maybe understand why?

We believe it may be because of the 32-bit architecture, we did not plan to release CAISAR on those platforms and thus we don't test against those.

Previous opam CI fails were because of an error on our side for packaging caisar-xgboost, we believe our last push fixed that. Multiple CI failures state that there is a HEAD.lock file, which we don't think comes from our side?

File: packages/caisar-onnx/caisar-onnx.0.2/opam [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+  "ocplib-endian" {>= "1.0"}
+  "caisar-ir" {= version}
+  "odoc" {with-doc}
+]
+build: [
```
Comment (mseri) [inline packages/caisar-onnx/caisar-onnx.0.2/opam:21]:
```suggestion
conflicts: [ "result" {< "1.5"} ]
build: [
```
This should solve the `Result.t` not found error

**Comment by mseri** [comment]
There is an extra lower bound failure with “missing ocaml-protoc-plugin” that I don’t think depends on your packages and can be ignored for the time being

**Comment by mseri** [comment]
There is still some work to do:
```

#=== ERROR while compiling multicont.1.0.1 ====================================#
# context              2.2.0~alpha~dev | linux/x86_64 | ocaml-base-compiler.5.0.0 | pinned(https://github.com/dhil/ocaml-multicont/releases/download/v1.0.1/multicont-1.0.1.tbz)
# path                 ~/.opam/5.0/.opam-switch/build/multicont.1.0.1
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p multicont -j 31 @install
# exit-code            1
# env-file             ~/.opam/log/multicont-7-17bd90.env
# output-file          ~/.opam/log/multicont-7-17bd90.out
### output ###
# File "lib/dune", line 1, characters 0-319:
#  1 | (library
#  2 |  (name multicont)
#  3 |  (public_name multicont)
# ....
# 12 |   (mode native)
# 13 |   (flags :standard (:include c_native_flags.sexp))
# 14 |   (names multicont_stubs)))
# (cd _build/default && /home/opam/.opam/5.0/bin/ocamlopt.opt -w -40 -g -shared -linkall -I lib -o lib/multicont.cmxs lib/multicont.cmxa)
# /usr/bin/ld: cannot find -lmulticont_stubs_native: No such file or directory
# collect2: error: ld returned 1 exit status
# File "caml_startup", line 1:
# Error: Error during linking (exit code 1)
```
And when only the bytecode is available
```
#=== ERROR while compiling multicont.1.0.1 ====================================#
# context              2.2.0~alpha~dev | linux/x86_32 | ocaml-base-compiler.5.0.0 | pinned(https://github.com/dhil/ocaml-multicont/releases/download/v1.0.1/multicont-1.0.1.tbz)
# path                 ~/.opam/5.0/.opam-switch/build/multicont.1.0.1
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p multicont -j 31 @install
# exit-code            1
# env-file             ~/.opam/log/multicont-7-440a72.env
# output-file          ~/.opam/log/multicont-7-440a72.out
### output ###
# File "lib/dune", line 4, characters 13-19:
# 4 |  (modes byte native)
#                  ^^^^^^
# Error: this mode isn't available
```

File: packages/caisar-onnx/caisar-onnx.0.2/opam (hunk 5 above)
Comment (mseri) [inline packages/caisar-onnx/caisar-onnx.0.2/opam:21]:
```suggestion
conflicts: [ "result" {< "1.5"} ]
build: [
```

**Comment by mseri** [comment]
Thanks! Feel free to announce the release on discuss.ocaml.org. We have a community category and an announcement tag for this purpose
